AutoCollection/HttpDependencies.ts

```diff
@@ -55,29 +56,43 @@ class AutoCollectHttpDependencies {
         this._isInitialized = true;
 
         const originalRequest = http.request;
+        const originalHttpsRequest = https.request;
+
+        // On node >= v0.11.12 and < 9.0 (excluding 8.9.0) https.request just calls http.request (with additional options).
+        // On node < 0.11.12, 8.9.0, and 9.0 > https.request is handled separately
+        // Patch both and leave a flag to not double-count on versions that just call through
+        // We add the flag to both http and https to protect against strange double collection in other scenarios
         http.request = (options, ...requestArgs: any[]) => {
             const request: http.ClientRequest = originalRequest.call(
                 http, options, ...requestArgs);
-            if (request && options && !(<any>options)[AutoCollectHttpDependencies.disableCollectionRequestOption]) {
+
+            var shouldCollect = !(<any>options)[AutoCollectHttpDependencies.disableCollectionRequestOption] &&
+                !(<any>request)[AutoCollectHttpDependencies.alreadyAutoCollectedFlag];
+
+            (<any>request)[AutoCollectHttpDependencies.alreadyAutoCollectedFlag] = true;
+
+            if (request && options && shouldCollect) {
                 AutoCollectHttpDependencies.trackRequest(this._client, options, request);
             }
+
             return request;
         };
 
-        // On node >= v0.11.12, https.request just calls http.request (with additional options).
-        // But on older versions, https.request needs to be patched also.
-        // The regex matches versions < 0.11.12 (avoiding a semver package dependency).
-        if (/^0\.([0-9]\.)|(10\.)|(11\.([0-9]|10|11)$)/.test(process.versions.node)) {
-            const originalHttpsRequest = https.request;
-            https.request = (options, ...requestArgs: any[]) => {
-                const request: http.ClientRequest = originalHttpsRequest.call(
-                    https, options, ...requestArgs);
-                if (request && options && !(<any>options)[AutoCollectHttpDependencies.disableCollectionRequestOption]) {
-                    AutoCollectHttpDependencies.trackRequest(this._client, options, request);
-                }
-                return request;
-            };
-        }
+        https.request = (options, ...requestArgs: any[]) => {
+            const request: http.ClientRequest = originalHttpsRequest.call(
```

Comment:
isn't it exactly the same as for http.request? Would it make sense to move it to a method and just pass originalHttpsRequest/originalRequest object?

Reply:
Nope, the base functions take different numbers of arguments 😄

Reply:
I do really dislike the duplication though..

Reply:
Ok, let's leave it as it is.

Reply:
Actually scratch that, I think this can be cleanly de-duplicated. Will change